src/github.com/matrix-org/go-neb/services/slackapi/message.go

```diff
@@ -0,0 +1,219 @@
+	AuthorLink    template.URL `json:"author_link"`
+	AuthorIcon    *string      `json:"author_icon"`
+	AuthorIconURL template.URL
+
+	Title     *string `json:"title"`
+	TitleLink *string `json:"title_link"`
+
+	Text         string `json:"text"`
+	TextRendered template.HTML
+
+	MrkdwnIn []string `json:"mrkdwn_in"`
+	Ts       *int64   `json:"ts"`
+}
+
+type slackMessage struct {
+	Text         string `json:"text"`
+	TextRendered template.HTML
+	Username     string            `json:"username"`
+	Channel      string            `json:"channel"`
+	Mrkdwn       *bool             `json:"mrkdwn"`
+	Attachments  []slackAttachment `json:"attachments"`
+}
+
+// We use text.template because any fields of any attachments could
+// be Markdown, so it's convenient to escape on a field-by field basis.
+// We do not do this yet, since it's assumed that clients also escape the content we send them.
+var htmlTemplate, _ = template.New("htmlTemplate").Parse(`
+<strong>@{{ .Username }}</strong> via <strong>#{{ .Channel }}</strong><br />
+{{- with (or .TextRendered .Text nil) }}
+	{{- if . }}
+		{{- . }}<br />
+	{{- end }}
+{{- end }}
+{{- range .Attachments }}
+		{{- if .AuthorName }}
+			{{- if .AuthorLink }}<a href="{{ .AuthorLink }}">{{ end }}
+				{{- if .AuthorIconUrl }}<img src="{{ .AuthorIconUrl }}" />{{ end }}
+				{{- .AuthorName }}
+			{{- if .AuthorLink }}</a>{{ end }}
+			<br />
+		{{- end }}
+	<strong>
+		<font color="{{- .ColorRendered }}">▌</font>
+		{{- if .TitleLink }}
+			<a href="{{ .TitleLink}}">{{ .Title }}</a>
+		{{- else }}
+			{{- .Title }}
+		{{- end }}
+		<br />
+	</strong>
+	{{- if .Pretext }}{{ or .PretextRendered .Pretext }}<br />{{ end }}
+	{{- if .Text }}{{ or .TextRendered .Text }}<br />{{ end }}
+{{- end }}
+`)
+
+var netClient = &http.Client{
+	Timeout: time.Second * 10,
+}
+
+var linkRegex, _ = regexp.Compile("<([^|]+)(\\|([^>]+))?>")
+
+func getSlackMessage(req http.Request) (message slackMessage, err error) {
+	ct := req.Header.Get("Content-Type")
+	ct, _, err = mime.ParseMediaType(ct)
+
+	if ct == "application/x-www-form-urlencoded" {
+		req.ParseForm()
+		payload := req.Form.Get("payload")
+		err = json.Unmarshal([]byte(payload), &message)
+	} else if ct == "application/json" {
+		log.Info("Parsing as JSON")
+		decoder := json.NewDecoder(req.Body)
+		err = decoder.Decode(&message)
+	} else {
+		message.Text = fmt.Sprintf("**Error:** unknown Content-Type `%s`", ct)
+		log.Error(message.Text)
+	}
+
+	return
+}
+
+func linkifyString(text string) string {
+	return linkRegex.ReplaceAllString(text, "<a href=\"$1\">$3</a>")
+}
+
+func getColor(color *string) string {
+	if color == nil {
+		return "black"
+	}
+
+	// https://api.slack.com/docs/message-attachments defines these aliases
+	mappedColor, ok := map[string]string{
+		"good":    "green",
+		"warning": "yellow",
+		"danger":  "red",
+	}[*color]
+	if ok {
+		return mappedColor
+	}
+	return *color
+}
+
+// fetches an image and encodes it as a data URL
+// returns nil if fetch fails
```

Comment:
No it doesn't :angel: It returns a 0-length string. Named return args are default-initialised and `template.URL` is an alias for a `string`.